### PR TITLE
fix: cover openai and gemini endpoint surfaces

### PR DESCRIPTION
**Pull request by ding113**
## Summary
Introduce a shared endpoint family catalog that centralizes endpoint classification for Claude, Response API, OpenAI, Gemini, and Gemini CLI surfaces. This classifies official OpenAI resource endpoints and additional Gemini model actions as standard proxy traffic instead of MCP passthrough, and gates Gemini Google Search provider overrides to generation-only actions.

## Problem
The codebase had endpoint classification logic duplicated across multiple files:
1. `format-mapper.ts` had inline regex patterns for endpoint format detection
2. `forwarder.ts` maintained its own `STANDARD_ENDPOINTS` and `STRICT_STANDARD_ENDPOINTS` arrays that diverged from `endpoint-paths.ts`
3. Gemini Google Search provider overrides were applied to all Gemini endpoints, including non-generation endpoints like `countTokens` and `embedContent`

This was noted in PR #931's review: "Duplicate endpoint lists risk drift - these two sources have already diverged."

**Related PRs:**
- Follow-up to #931 - Consolidates endpoint classification after embeddings passthrough addition
- Builds on #801 - Extends EndpointPolicy architecture with centralized endpoint catalog

## Solution
1. Create `endpoint-family-catalog.ts` with a declarative `EndpointFamily` interface that captures surface type, accounting tier, model requirement, and passthrough behavior
2. Replace scattered endpoint arrays in `forwarder.ts` with calls to `isStandardProxyEndpointPath()` from the catalog
3. Delegate format detection in `format-mapper.ts` to `detectEndpointFormat()` from the catalog
4. Gate Gemini Google Search overrides to generation-only endpoints (`generatecontent`, `streamgeneratecontent`)

### Core Changes
- `src/app/v1/_lib/proxy/endpoint-family-catalog.ts` (+443) - New centralized endpoint catalog with 25+ endpoint families
- `src/app/v1/_lib/proxy/format-mapper.ts` (+32/-46) - Delegate to catalog, add Gemini batch format detection
- `src/app/v1/_lib/proxy/forwarder.ts` (+12/-24) - Replace inline endpoint arrays with catalog functions
- `src/lib/gemini/provider-overrides.ts` (+14/-3) - Gate Google Search override to generation endpoints

### Supporting Changes
- `src/app/v1/_lib/proxy/provider-selector.ts` (+4/-3) - Minor import adjustment

### Tests
- `tests/unit/proxy/endpoint-family-catalog.test.ts` (+372) - Comprehensive catalog coverage
- `tests/unit/proxy/openai-official-standard-forwarder.test.ts` (+144) - OpenAI standard endpoint routing
- `tests/unit/proxy/provider-selector-resource-endpoints.test.ts` (+112) - Resource endpoint selection
- `tests/unit/lib/gemini/provider-overrides.test.ts` (+32) - Google Search override gating
- `tests/unit/proxy/extract-usage-metrics.test.ts` (+38) - Usage extraction edge cases

### Automated Tests
- [x] Unit tests added for endpoint family catalog
- [x] Unit tests added for OpenAI standard forwarder routing
- [x] Unit tests added for provider selector resource endpoints
- [x] Unit tests added for Gemini Google Search override gating

### Verification Commands
```bash
bun run build
bun run lint
bun run lint:fix
bun run typecheck
bun run test
bunx vitest run tests/unit/proxy tests/unit/lib/gemini/provider-overrides.test.ts tests/unit/app/v1/url.test.ts
```

## Breaking Changes
None. This is an internal refactoring that maintains backward compatibility.

## Checklist
- [x] Code follows project conventions
- [x] Self-review completed
- [x] Tests pass locally
- [x] No breaking changes

---
*Description enhanced by Claude AI*

<details open><summary><h3>Greptile Summary</h3></summary>

This PR consolidates scattered endpoint-classification logic into a single `endpoint-family-catalog.ts` module, replacing two separate hard-coded arrays in `forwarder.ts`, ad-hoc inline regexes in `format-mapper.ts`, and an over-broad Google Search override application in `provider-overrides.ts`. The refactor correctly classifies ~35 previously-MCP-passthrough OpenAI resource endpoints and additional Gemini model actions as standard proxy traffic, and gates the Google Search tool injection/removal to generation-only Gemini endpoints.

Key changes:
- **`endpoint-family-catalog.ts`** — new declarative catalog of 42 `EndpointFamily` entries covering Claude, Response, OpenAI, Gemini, and Gemini-CLI surfaces; exports `resolveEndpointFamilyByPath`, `isStandardProxyEndpointPath`, `detectEndpointFormat`, and `isGeminiGenerationEndpointPath`.
- **`forwarder.ts`** — removes the `STANDARD_ENDPOINTS` / `STRICT_STANDARD_ENDPOINTS` arrays and delegates to catalog functions; passes `session.requestUrl.pathname` into the Google Search override audit.
- **`provider-selector.ts`** — changes the no-model fallback from "claude-only" to "any provider", safe because format-type compatibility is already enforced one step earlier (step 2b).
- **`provider-overrides.ts`** — gates Google Search overrides to paths where `isGeminiGenerationEndpointPath` returns `true`, preventing injection into `countTokens`/`embedContent` calls.
- The `GEMINI_GENERATION_ACTIONS` Set in `endpoint-family-catalog.ts` is a second source of truth that partially duplicates the generation-family definitions already present in the catalog, risking future drift.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

- Safe to merge; behavioral changes are intentional and well-covered by new tests. Minor style concerns only.
- The core logic is sound: first-match-wins ordering in the catalog is correct for all overlapping matchers (response-compact before response-resources, exact endpoint entries before prefix entries), and the provider-selector change is guarded by the format-compatibility filter that runs earlier in the same function. The main residual risk is that `GEMINI_GENERATION_ACTIONS` is a second source of truth for generation action names that could drift from the catalog, but this does not affect current correctness.
- Pay close attention to `src/app/v1/_lib/proxy/endpoint-family-catalog.ts` — specifically the `GEMINI_GENERATION_ACTIONS` Set and `isGeminiGenerationEndpointPath` which maintain independent knowledge of which actions are "generation" endpoints.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/proxy/endpoint-family-catalog.ts | New centralized catalog of 42 endpoint families. Core logic is correct; minor issues: GEMINI_GENERATION_ACTIONS is a second source of truth that can drift from the catalog, /i regex flags are redundant after path normalization, and openai-completions uses an overly broad hasPrefix matcher. |
| src/app/v1/_lib/proxy/forwarder.ts | Replaces two hard-coded endpoint arrays with catalog-backed isStandardProxyEndpointPath / isStrictStandardEndpointPath, and passes the request pathname through to the Google Search override audit. Straightforward and low-risk. |
| src/app/v1/_lib/proxy/provider-selector.ts | Changes model-missing fallback from "claude providers only" to "any provider". Safe because the format-type filter (step 2b) already enforces provider-type compatibility before the model check is reached, as confirmed by the new resource-endpoint tests. |
| src/lib/gemini/provider-overrides.ts | Gates Google Search tool injection/removal to generation-only Gemini endpoints; non-generation paths and absent/null pathnames are handled correctly with early returns. |
| tests/unit/proxy/endpoint-family-catalog.test.ts | Good coverage of all 42 families; the Set-based coverage assertion is weakened by three duplicate gemini-batch-embed-contents entries that collapse to one ID in the Set comparison. |
| tests/unit/proxy/provider-selector-resource-endpoints.test.ts | Correctly validates that openai/response/gemini resource endpoints select the right provider type even without a model; provides solid regression coverage for the provider-selector change. |

</details>

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Incoming Request\npathname] --> B[normalizeEndpointPath]
    B --> C[resolveEndpointFamilyByPath\nfind first matching family]

    C --> D{Family found?}
    D -- No --> E[null → MCP passthrough\n unless gemini/gemini-cli provider]
    D -- Yes --> F[EndpointFamily\nid · surface · accountingTier\nmodelRequired · rawPassthrough]

    F --> G[isStandardProxyEndpointPath\nforwarder: isMcpRequest=false]
    F --> H[detectEndpointFormat\nformat-mapper: surface type]
    F --> I[isStrictStandardEndpointPath\nforwarder: enforce strict pool]

    A --> J[isGeminiGenerationEndpointPath\nregex extract :action]
    J --> K{action in\nGEMINI_GENERATION_ACTIONS?}
    K -- Yes --> L[Apply Google Search\ntool override]
    K -- No --> M[Skip override\ncountTokens / embedContent]

    style E fill:#f9c,stroke:#c66
    style L fill:#cfc,stroke:#6c6
    style M fill:#eee,stroke:#999
```
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/app/v1/_lib/proxy/endpoint-family-catalog.ts
Line: 16

Comment:
**`GEMINI_GENERATION_ACTIONS` duplicates catalog knowledge**

`GEMINI_GENERATION_ACTIONS` on line 16 is a second, independent source of truth that shadows the generation-family definitions already declared in `KNOWN_ENDPOINT_FAMILIES`. If a future generation action (e.g., `livegeneratecontent`) is added to the catalog as a new family, `isGeminiGenerationEndpointPath` will silently miss it unless this Set is updated in tandem — exactly the drift problem this PR was introduced to fix.

The `rawPassthrough`/`accountingTier` fields already live on `EndpointFamily`. Adding a boolean `isGenerationAction: boolean` (or a `role: "generation" | "embedding" | "resource"` discriminant) to the interface would let `isGeminiGenerationEndpointPath` derive its answer entirely from the catalog:

```ts
export function isGeminiGenerationEndpointPath(pathname: string): boolean {
  const family = resolveEndpointFamilyByPath(pathname);
  return (
    family !== null &&
    (family.surface === "gemini" || family.surface === "gemini-cli") &&
    family.isGenerationAction === true
  );
}
```

This keeps the catalog as the single source of truth and removes the risk of the two lists diverging again.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/app/v1/_lib/proxy/endpoint-family-catalog.ts
Line: 143

Comment:
**Redundant `/i` flag on already-lowercased path**

`normalizeEndpointPath` (imported on line 1, called on every lookup in `resolveEndpointFamilyByPath`) lowercases the pathname before it reaches any `match` function. The `/i` flag in this regex — and in the two regexes inside the `gemini-models-resource` matcher at lines 254–255 — is therefore redundant. It adds no correctness benefit and creates a minor misleading impression that paths may arrive in mixed case at this point.

```suggestion
    match: (pathname) => /^\/v1\/models(?:\/[^/:]+)?$/.test(pathname),
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/app/v1/_lib/proxy/endpoint-family-catalog.ts
Line: 282-283

Comment:
**`openai-completions` over-matches with `hasPrefix`**

The legacy completions endpoint is a single action endpoint (`POST /v1/completions`) — there are no sub-resources under it. Using `hasPrefix(pathname, "/v1/completions")` means any hypothetical path like `/v1/completions/cmpl_123` would be matched as an `openai-completions` family with `modelRequired: true` and `accountingTier: "required_usage"`. A precise exact match would be more consistent with how `openai-embeddings` and the generation endpoints are handled:

```suggestion
    match: (pathname) => pathname === "/v1/completions",
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: tests/unit/proxy/endpoint-family-catalog.test.ts
Line: 291-303

Comment:
**Duplicate `id` entries make the coverage assertion weaker than it looks**

`FAMILY_SAMPLES` includes three entries with `id: "gemini-batch-embed-contents"` (lines 249–255, 291–296, and 297–303). The "样例应覆盖所有已知端点族" test at line 321 uses `new Set(...)`, which silently collapses duplicates — so the equality check would still pass even if one of those three sample paths resolved to the wrong family, as long as the ID appears at least once.

The two extra entries are presumably testing the `/v1/publishers/google/models/` and `/v1/models/` prefix variants of the same family, which is a valuable scenario. Consider giving them distinct labels or a comment explaining they intentionally test alternate prefixes for the same family ID, so a future reader doesn't accidentally remove them as "duplicates".

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: 4388404</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

File: src/app/v1/_lib/proxy/format-mapper.ts [hunk 1]
```diff
@@ -121,13 +82,38 @@ export function detectClientFormat(requestBody: Record<string, unknown>): Client
...
+        return typeof nestedRequest.model === "string" || typeof nestedRequest.content === "object";
+      }
+
+      return false;
+    });
+
+    if (hasGeminiBatchShape) {
+      return "gemini";
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/format-mapper.ts:119]: 已收紧 Gemini batch body 检测：现在要求 requests 条目满足 model + 有效的 Gemini content payload（非 null、非数组，且包含 parts/contents），避免把宽泛 object 误判成 gemini。对应定向与全量测试已通过。

File: src/app/v1/_lib/proxy/endpoint-family-catalog.ts [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+  },
+  {
+    id: "gemini-cli-stream-generate-content",
+    surface: "gemini-cli",
+    accountingTier: "required_usage",
+    modelRequired: true,
+    rawPassthrough: false,
+    match: (pathname) =>
+      matchGeminiModelAction(pathname, "/v1internal/models/", ["streamgeneratecontent"]),
+  },
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/endpoint-family-catalog.ts:276]: 已做这块收敛：提取了 Gemini 前缀集合和统一的前缀匹配 helper，catalog 里不再手写重复的多段 OR 匹配。

File: src/app/v1/_lib/proxy/endpoint-family-catalog.ts [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+
+function normalizePathname(pathname: string): string {
+  const pathWithoutQuery = pathname.split("?")[0] ?? pathname;
+  const trimmedPath =
+    pathWithoutQuery.length > 1 && pathWithoutQuery.endsWith("/")
+      ? pathWithoutQuery.slice(0, -1)
+      : pathWithoutQuery;
+
+  return trimmedPath.toLowerCase();
+}
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/endpoint-family-catalog.ts]: 已复用现有 normalizeEndpointPath，catalog 内不再维护一份重复的路径归一化逻辑。

File: src/app/v1/_lib/proxy/endpoint-family-catalog.ts [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    | "/v1/models/"
+    | "/v1internal/models/",
+  actions: readonly string[]
+): boolean {
+  const actionPattern = actions.join("|");
+  const regex = new RegExp(
+    `^${prefix.replaceAll("/", String.raw`\/`)}[^/:]+:(?:${actionPattern})$`,
+    "i"
+  );
+  return regex.test(pathname);
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/endpoint-family-catalog.ts]: 已改为字符串解析路径，不再在热路径里每次动态编译 RegExp。现在是按 prefix + model/action 分段判断。
